crates/core/tedge_agent/src/state_repository/state.rs

```diff
@@ -1,186 +1,106 @@
 use crate::state_repository::error::StateError;
-use async_trait::async_trait;
 use camino::Utf8PathBuf;
-use serde::Deserialize;
+use serde::de::DeserializeOwned;
 use serde::Serialize;
-use std::str::FromStr;
+use std::marker::PhantomData;
 use tedge_utils::fs::atomically_write_file_async;
 use tokio::fs;
 
+/// Store the current state of an operation
 #[derive(Debug)]
-pub struct AgentStateRepository {
+pub struct AgentStateRepository<T> {
```

Comment:
Just a remark: Now that this struct is used to store the state at a per operation level and not as a repository of states of multiple operations, the `Repository` in its name sounds a bit misleading. But, we can live with that for now.

Reply:
This struct is used to store the state at a per operation level since the restart operation has been introduced. However, this were really weird as each operation were ready to have in its store foreign data rejected with unknown operation errors.